src/net/server/handlers/message_stream_layer/service.rs

```diff
@@ -0,0 +1,134 @@
+use super::future::ResponseFuture;
+use crate::net::server::MpcServerError;
+use axum::http::{Request, Response};
+use std::task::{ready, Context, Poll};
+use tokio::sync::mpsc;
+use tokio_util::sync::PollSender;
+use tower::Service;
+
+#[derive(Debug)]
+pub enum SenderStatus<T> {
+    NotReady(PollSender<T>),
+    Ready(PollSender<T>),
+    Err,
+    Spent,
+}
+
+// manually implement Clone to get around requirements that `T` must be [`Clone`]
+impl<T> Clone for SenderStatus<T> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::NotReady(sender) => Self::NotReady(sender.clone()),
+            Self::Ready(sender) => Self::Ready(sender.clone()),
+            Self::Err => Self::Err,
+            Self::Spent => Self::Spent,
+        }
+    }
+}
+
+pub struct ReservedPermit<T>(PollSender<T>);
+impl<T: Send + 'static> ReservedPermit<T> {
+    pub fn send(&mut self, item: T) -> Result<(), MpcServerError> {
+        Ok(self.0.send_item(item)?)
+    }
+}
+
+/// `poll_ready` will first grab a permit via the `PollSender`. Once ready, the sender is injected
```

Comment:
lets work out a better name and description for it

Reply:
code is now obsolete